cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,710 @@
+
+  // get_number_of_local_adj_matrix_partition_rows == summation of get_major_size() of all
+  // partitions belonging to the gpu
+  vertex_t partial_offset{0};
+  for (size_t i = 0; i < graph_view.get_number_of_local_adj_matrix_partitions(); ++i) {
+    auto matrix_partition =
+      matrix_partition_device_view_t<vertex_t, edge_t, weight_t, GraphViewType::is_multi_gpu>(
+        graph_view.get_matrix_partition_view(i));
+
+    // Check if hypersparse segment is present in the partition
+    auto segment_offsets = graph_view.get_local_adj_matrix_partition_segment_offsets(i);
+    auto use_dcs =
+      segment_offsets
+        ? ((*segment_offsets).size() > (detail::num_sparse_segments_per_vertex_partition + 1))
+        : false;
+
+    if (use_dcs) {
+      auto major_hypersparse_first =
+        matrix_partition.get_major_first() +
+        (*segment_offsets)[detail::num_sparse_segments_per_vertex_partition];
+      // Calculate degrees in csr region
+      auto csr_begin = local_degrees.begin() + partial_offset;
+      auto csr_end   = local_degrees.begin() + partial_offset +
+                     (major_hypersparse_first - matrix_partition.get_major_first());
+      ;
+
+      thrust::tabulate(handle.get_thrust_policy(),
+                       csr_begin,
+                       csr_end,
+                       [offsets = matrix_partition.get_offsets()] __device__(auto i) {
+                         return offsets[i + 1] - offsets[i];
+                       });
+
+      // Calculate degrees in dcsr region
+      auto dcs_nzd_vertex_count = *(matrix_partition.get_dcs_nzd_vertex_count());
+      // Initialize hypersparse region degrees as 0
+      thrust::fill(handle.get_thrust_policy(),
+                   csr_end,
+                   csr_begin + matrix_partition.get_major_size(),
+                   edge_t{0});
+      thrust::for_each(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator(vertex_t{0}),
+                       thrust::make_counting_iterator(dcs_nzd_vertex_count),
+                       [major_hypersparse_first,
+                        major_first   = matrix_partition.get_major_first(),
+                        vertex_ids    = *(matrix_partition.get_dcs_nzd_vertices()),
+                        offsets       = matrix_partition.get_offsets(),
+                        local_degrees = thrust::raw_pointer_cast(csr_begin)] __device__(auto i) {
+                         auto d = offsets[(major_hypersparse_first - major_first) + i + 1] -
+                                  offsets[(major_hypersparse_first - major_first) + i];
+                         auto v                         = vertex_ids[i];
+                         local_degrees[v - major_first] = d;
+                       });
+    } else {
+      auto csr_begin = local_degrees.begin() + partial_offset;
+      auto csr_end   = local_degrees.begin() + partial_offset + matrix_partition.get_major_size();
```

Comment:
If we compute local major degrees, this is no longer necessarily a CSR but can be either CSR or CSC. Maybe use sparse_begin() and sparse_end() instead?